In [1]:
import numpy as np

**1) Как относятся друг к другу множество и последовательность? (в ответе использовать слова типа: часть, целое, общее, частное, родитель, дочерний субъект и т.д.)?**

Последовательность это функция аргументом которой является элемент из множества натуральных чисел. Значения этой функции принадлежат некоторому множеству, которое называется областью значения функции.

**2) Прочитать высказывания математической логики, построить их отрицания и *установить истинность.***


$\forall y \in[0 ; 1] : \operatorname{sgn}(y)=1$

Для любого y, принадлежащего отрезку от 0 до 1 включительно, верно, что значение функции sgn(y) равно 1.  
**Это выражение ложно.**  

Отрицание:  
$\exists y \in[0 ; 1] : \operatorname{sgn}(y)\not=1$

$\forall n \in \mathbb{N}>2 : \exists x, y, z \in \mathbb{N} : x^{n}=y^{n}+z^{n}$

Для любого n из множества натуральных чисел и больше 2 существуют такие x,y,z из множества натуральных чисел, что справедливо равенство x^n = y^n + z^n  
**Это выражение ложно.**  

Отрицание:  
$\exists n \in \mathbb{N}>2 : \forall x, y, z \in \mathbb{N} : x^{n}\not=y^{n}+z^{n}$


$\forall x \in \mathbb{R} \exists X \in \mathbb{R} : X>x$

Для любого x из множества действительных чисел существует такое X из множества действительных чисел, что X > x.  
**Это выражение истинно.**  

Отрицание:  
$\exists x \in \mathbb{R} \forall X \in \mathbb{R} : X\leq x$


$\forall x \in \mathbb{C} \not \exists y \in \mathbb{C} : x>y \| x<y$

Для любого х из множества комплексных чисел не существует такого y из множества комплексных чисел что справедливо неравенство x > y или x < y  
**Это выражение ложно, так как нельзя сравнивать комплексные числа на отношение больше/меньше.**  

Отрицание:  
$\exists x \in \mathbb{C} \exists y \in \mathbb{C} : x \leq y , x \geq y$

$\forall y \in\left[0 ; \frac{\pi}{2}\right] \exists \varepsilon>0 : \sin y<\sin (y+\varepsilon)$

Для любого y, принадлежащего отрезку от 0 до $\frac{\pi}2$ включительно, существует такое положительное значение $\varepsilon$, что справедливо неравенство sin(y) < sin(y+$\varepsilon$)  
**Это выражение ложно.**  

Отрицание:  
$\exists y \in\left[0 ; \frac{\pi}{2}\right] \forall \varepsilon>0 : \sin y \geq \sin (y+\varepsilon)$

$\forall y \in[0 ; \pi) \exists \varepsilon>0 : \cos y>\cos (y+\varepsilon)$

Для любого y, принадлежащего отрезку от 0 включительно до ${\pi}$ ислючительно, существует такое положительное значение $\varepsilon$, что справедливо неравенство cos(y) > cos(y+$\varepsilon$)  
**Это выражение ложно.**  

Отрицание:  
$\exists y \in[0 ; \pi) \forall \varepsilon>0 : \cos y \leq \cos (y+\varepsilon)$


$\exists x : x \notin\{\mathbb{N}, \mathbb{Z}, \mathbb{Q}, \mathbb{R}, \mathbb{C}\}$

Для любого х справедливо, что х принадлежит множеству натуральных или целых или рациональных или вещественных или комплексных чисел.  
**Это варажение ложно.**  
Отрицание:  
$\forall x : x \in\{\mathbb{C}\}$


**3.1) Даны три множества a,b и с. Необходимо выполнить все изученные виды бинарных операций над всеми комбинациями множеств.**

###### Памятка для себя
**Бинарные операции**:  
1) Пересечение  
2) Объединение  
3) Разность  
4) Симметрическая разность  
5) Декартово произведение  

**Унарные операции**:  
1) Дополнение  
2) Булеан

**Порядок выполнения:**  
В первую очередь выполняются унарные операции, затем пересечение, затем все остальные (слева направо), имеющие равный приоритет.

In [2]:
A = set([10,20,30,40])
B = set([20,40,60,80,100])
C = set([10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110])

In [3]:
C - B

{10, 30, 50, 70, 90, 110}

In [4]:
C - B - A

{50, 70, 90, 110}

**3.2) *Выполнить задание 3.1 на языке Python**

**5.1)  *На языке Python предложить алгоритм вычисляющий численно предел $$\lim _{n \rightarrow+\infty} \frac{n}{\sqrt[n]{n !}}$$ с точностью $\varepsilon=10^{-7}$**


##### формула Стирлинга (также формула Муавра — Стирлинга) для аппроксимации функции факториала  
$$n ! \approx \sqrt{2 \pi n}\left(\frac{n}{e}\right)^{n}$$  

После подстановки формулы Стирлинга в функцию получаем:  
$$\lim _{n \rightarrow+\infty} \frac{e}{\sqrt[2n]{\pi n}}$$

In [5]:
def f(n):
    return (np.exp(1))/((np.pi*n)**(1/(2*n)))

Так как значения функции Стирлинга имеют приближенное значение к значению функции факториала, с разницей, ориентировочно, в 1 порядок, будем искать значение такие два ближайшие значения функции Стрилинга, разность которых равна 0.00000001

In [6]:
prev = f(1)
for i in range(2, 200000000):
    new_val = f(i)
    if new_val - prev > 0.00000001:
        prev = new_val
    else:
        print(i, new_val, prev)
        break

38121 2.717864957011133 2.7178649470113654


За 38121 итерацию установлено что с точностью $\varepsilon=10^{-7}$ значение предела функции равно 2.717864957011133

**5.2) *Предложить оптимизацию алгоритма, полученного в задании 5.1, ускоряющую его сходимость.**

Для уменьшения количества итераций можно ускорить перебор значений n

In [7]:
i = 2
iters_num = 0
while True:
    iters_num += 1
    prev = f(i*2)
    new_val = f(i)
    dif_sign = np.log10(abs(new_val - prev))
#     print(dif_sign)
    if dif_sign > -8:
        i *= 10
    else:
        print(i, new_val, prev)
        print("Iters num:", iters_num)
        break

2000000000 2.7182818131271587 2.7182818205575807
Iters num: 10


За десять итераций и двадцать просчетов функции Стирлинга мы получили более точное значение предела функции